### PR TITLE
chore: single batch evaluation for local test mode [DET-2931]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- supporting `PyTorch`, `TFEstimator` and `TFKeras` trials

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/_execution.py [hunk 1]
```diff
@@ -65,6 +65,7 @@ def _make_local_execution_env(
     config: Optional[Dict[str, Any]],
     hparams: Optional[Dict[str, Any]] = None,
     limit_gpus: Optional[int] = None,
+    test_mode: bool = False,
```
Comment (rb-determined-ai) [inline harness/determined/_execution.py:68]: Can we make this a required parameter?  It's not really optional; we always need to know if the environment is in test mode, the same as we always need to know if it's a managed_training or not.

File: harness/determined/experimental/_native.py [hunk 2]
```diff
@@ -97,7 +97,7 @@ def _make_test_workloads(
     check.eq(len(batch_metrics), config.scheduling_unit())
     logging.debug(f"Finished training, metrics: {batch_metrics}")
```
Comment (rb-determined-ai) [inline harness/determined/experimental/_native.py:98]: Can you also make this logging.debug into an info while you are here?  Also below with the validation metrics.

File: harness/determined/experimental/_native.py [hunk 3]
```diff
@@ -192,13 +192,16 @@ def test_one_batch(
     trial_class: Optional[Type[det.Trial]] = None,
     config: Optional[Dict[str, Any]] = None,
 ) -> Any:
+
```
Comment (rb-determined-ai) [inline harness/determined/experimental/_native.py:195]: this extra line seems unnecessary

File: harness/determined/pytorch/_pytorch_trial.py [hunk 4]
```diff
@@ -334,8 +335,10 @@ def _compute_validation_metrics(self) -> workload.Response:
             batch_metrics = []
 
             self.validation_loader = cast(torch.utils.data.DataLoader, self.validation_loader)
-            check.gt(len(self.validation_loader), 0)
-            for batch in self.validation_loader:
+            val_loader_len = len(self.validation_loader)
+            check.gt(val_loader_len, 0)
+            batch_stop_idx = val_loader_len if not self.context.env.test_mode else 1
+            for batch in islice(self.validation_loader, 0, batch_stop_idx):
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:341]: non-blocking:  What if you just put an `if self.env.test_mode: break` at the end of the loop?  I had never heard of `islice()` before, and while it is definitely not wrong to use advanced features of a language, in this case the simple thing seems so simple it's at least worth suggesting.

File: harness/determined/keras/_tf_keras_trial.py [hunk 5]
```diff
@@ -747,6 +747,9 @@ def _launch_evaluate(self) -> Any:
         use_model_metrics = version.parse(tf.__version__) < version.parse("2.2.0")
         evaluate_kwargs = {} if use_model_metrics else {"return_dict": True}
 
+        if self.context.env.test_mode:
```
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:750]: throughout: All these TrialController objects have a `self.env` that you should use directly.

File: harness/determined/experimental/_native.py [hunk 6]
```diff
@@ -192,13 +192,16 @@ def test_one_batch(
...
     # Override the scheduling_unit value to 1.
     # TODO(DET-2931): Make the validation step a single batch as well.
     config = {**(config or {}), "scheduling_unit": 1}
 
     logging.info("Running a minimal test experiment locally")
     checkpoint_dir = tempfile.TemporaryDirectory()
-    env, rendezvous_info, hvd_config = det._make_local_execution_env(True, config, limit_gpus=1)
+    env, rendezvous_info, hvd_config = det._make_local_execution_env(
+        True, config, limit_gpus=1, test_mode=True
```
Comment (rb-determined-ai) [inline harness/determined/experimental/_native.py:203]: while you are here, can you switch this to be all keyword arguments?  That initial bare `True` is a little opaque.

**Review by rb-determined-ai — COMMENTED** [review]
This is looking great.

File: harness/determined/_execution.py (hunk 1 above)
Comment (hamidzr) [reply harness/determined/_execution.py:68]: hmm by providing default we'd make the usage simpler but it'll be less explicit. either way works

File: harness/determined/pytorch/_pytorch_trial.py (hunk 4 above)
Comment (hamidzr) [reply harness/determined/pytorch/_pytorch_trial.py]: originally I was designing it to allow arbitrary batch length but early breaking is simpler 👍🏼

**Review by rb-determined-ai — APPROVED** [review]
Looks fantastic.  Very nice work.

**Comment by hamidzr** [comment]
thanks @rb-determined-ai
